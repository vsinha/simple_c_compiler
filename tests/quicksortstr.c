

void printArray( char** a, long left, long right) {
	long i;
	for (i = left; i <= right; i = i+1) {
		printf("%d: %s\n", i, a[i]);
	}
}

void print(char* s) {
	printf("==%s==\n", s);
}

long mystrcmp(char* s1, char* s2) {
	while (s1[0] && s2[0] && s1[0] == s2[0]) {
		s1 = s1 + 1;
		s2 = s2 + 1;
	}

	if (s1[0] == 0 && s2[0] == 0) {
		return 0;
	}

	if (s1[0] == 0 && s2[0] != 0) {
		return -1;
	}
	
	if (s1[0] != 0 && s2[0] == 0) {
		return 1;
	}

	if (s1[0] > s2[0]) {
		return 1;
	}

	return -1;
	
}

long quicksortsubrange(char** a, long left, long right)
{
	if (left >= right) {
		return 0;
	}

	char* x;
	x = a[right];

	long l;
	l = left;
	long g;
	g = right - 1;

	while (l<g) {

		while (l<g && mystrcmp(a[l],x)<0) {
			l = l + 1;
		}

		while (l<g && mystrcmp(a[g],x)>=0) {
			g = g - 1;
		}

		if (l<g) {
			long tmp;
			tmp = a[l];
			a[l] = a[g];
			a[g]=tmp;
		}
	}

	a[right]=a[l];
	a[l]=x;

	quicksortsubrange(a, left, l - 1);
	quicksortsubrange(a,g+1, right);

}

void quicksort(long* a, long n) {
	quicksortsubrange(a, 0, n - 1);
}

void
main() {

	long n;
	n = 6;
	
	char** a;
	a = malloc(n*8);

	a[0] = "Rachael";
	a[1] = "Monica";
	a[2] = "Phoebe";
	a[3] = "Joey";
	a[4] = "Ross";
	a[5] = "Chandler";

	printf("-------- Before -------\n");
	printArray(a, 0, n - 1);
	quicksort(a, n);
	printf("-------- After -------\n");
	printArray(a, 0, n - 1);
}

