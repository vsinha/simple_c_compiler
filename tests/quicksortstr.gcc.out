quicksortstr.c: In function 'printArray':
quicksortstr.c:6:3: warning: incompatible implicit declaration of built-in function 'printf' [enabled by default]
quicksortstr.c:6:3: warning: format '%d' expects argument of type 'int', but argument 2 has type 'long int' [-Wformat]
quicksortstr.c: In function 'print':
quicksortstr.c:11:2: warning: incompatible implicit declaration of built-in function 'printf' [enabled by default]
quicksortstr.c: In function 'quicksortsubrange':
quicksortstr.c:66:8: warning: assignment makes integer from pointer without a cast [enabled by default]
quicksortstr.c:68:8: warning: assignment makes pointer from integer without a cast [enabled by default]
quicksortstr.c: In function 'quicksort':
quicksortstr.c:81:2: warning: passing argument 1 of 'quicksortsubrange' from incompatible pointer type [enabled by default]
quicksortstr.c:40:6: note: expected 'char **' but argument is of type 'long int *'
quicksortstr.c: In function 'main':
quicksortstr.c:91:6: warning: incompatible implicit declaration of built-in function 'malloc' [enabled by default]
quicksortstr.c:100:2: warning: incompatible implicit declaration of built-in function 'printf' [enabled by default]
quicksortstr.c:102:2: warning: passing argument 1 of 'quicksort' from incompatible pointer type [enabled by default]
quicksortstr.c:80:6: note: expected 'long int *' but argument is of type 'char **'
